core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -88,16 +92,26 @@ private ParallelIterator(
     @Override
     public void close() {
       // close first, avoid new task submit
-      this.closed = true;
+      this.closed.set(true);
+
+      try (Closer closer = Closer.create()) {
+        yieldedTasks.forEach(closer::register);
+        yieldedTasks.clear();
 
-      // cancel background tasks
-      for (Future<?> taskFuture : taskFutures) {
-        if (taskFuture != null && !taskFuture.isDone()) {
-          taskFuture.cancel(true);
+        // TODO close input iterables that were not started yet
```

Comment:
Is this needed? If this class hasn't created an iterator from an iterable, do we need to close the iterable? It's probably safer to but I'm not sure it is required.

Reply:
the existing code (in main branch) closes iterables, not the iterators derived from them
so to me it makes sense to reliably close all iterables.

i can back out this TODO though, if this is preferred.

Reply:
I think we should either get it done or decide not to, not just leave a TODO about closing here.

Aren't iterables added to a closing group and closed when the parallel iterator is closed?

Reply:
> The existing code use a try-with-resources for the iterable and used a `for` loop. 

I do believe that the code in this PR is semantically equivalent, so whatever assumptions/contract there were, they should not be violated by these changes.

> The way our iterables should work is that there is no need to close the iterable itself (i.e. no resources it holds) but it is a convenient way to close any iterators that were opened from the iterable, which do hold resources.

i find this mental model harder to reason about.
It's the iterable who is checked for being Closeable, so I would feel obliged to call close on the instance that was given to me, for me to own it.
But it's not the place to debate this model. I will remove the TODO. 

> We do need to make sure the yielded tasks or currently running tasks are closed. I think that closing the yielded tasks above and cancelling the running Task instances should take care of that. The running instances should hit the catch for Throwable that calls close.

Yes, mostly, but there is a race condition.
(a) a task may be about to yield or (b) already yielded but not moved to `yieldedTasks` collection yet.
In both cases `taskFuture.cancel(true)` won't cause the closing of the iterable in the task.

What the solution should be depends on whether the Task can be closed asyncly. If yes, it would makes things easier, but i don't think we can make such an assumption.

I will introduce timed-wait inside the close to wait for the tasks continuations and close them.

Reply:
> I will introduce timed-wait inside the close to wait for the tasks continuations and close them.

instead of time-wait, i switched the code to use CompletableFutures so that i can register closing callback, without adding more blocking within close().

Reply:
> It's the iterable who is checked for being Closeable, so I would feel obliged to call close on the instance that was given to me, for me to own it.

The purpose of this model is for cases where you use the iterable in a for loop:
```java
try {
  for (Item it : iterable) {
    ...
  }
} finally {
  // what should get closed here?
}
```

If you use that pattern, then there's nothing to close in the `finally` block because the iterator is not available. We added closable iterables to solve that. When you close an iterable, it closes all of the iterators that were opened from it.